From 6cb4adf815f25a42a386242253e2f81621ba0bc0 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Fri, 4 Apr 2014 15:26:17 +0100
Subject: [PATCH 179/351] ASoC: florida: Add support to sync up to a started
 ez2control stream

Add support in the trigger callback to check if we have already received
a trigger event before the stream was opened and manually process the
data from this. This will allow streams to be opened in response to the
trigger event if required.

Change-Id: If2ef7631175c5d0245766b5f33a6840cda5ca13b
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 sound/soc/codecs/florida.c |   11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/sound/soc/codecs/florida.c b/sound/soc/codecs/florida.c
index 80a01e2..01273aa 100644
--- a/sound/soc/codecs/florida.c
+++ b/sound/soc/codecs/florida.c
@@ -1909,12 +1909,20 @@ static int florida_trigger(struct snd_compr_stream *stream, int cmd)
 	struct snd_soc_pcm_runtime *rtd = stream->private_data;
 	struct florida_priv *florida = snd_soc_codec_get_drvdata(rtd->codec);
 	int ret = 0;
+	bool pending = false;
 
 	mutex_lock(&florida->compr_info.lock);
 
 	switch (cmd) {
 	case SNDRV_PCM_TRIGGER_START:
 		ret = wm_adsp_stream_start(florida->compr_info.adsp);
+
+		/**
+		 * If the stream has already triggered before the stream
+		 * opened better process any outstanding data
+		 */
+		if (florida->compr_info.trig)
+			pending = true;
 		break;
 	case SNDRV_PCM_TRIGGER_STOP:
 		break;
@@ -1925,6 +1933,9 @@ static int florida_trigger(struct snd_compr_stream *stream, int cmd)
 
 	mutex_unlock(&florida->compr_info.lock);
 
+	if (pending)
+		adsp2_irq(0, florida);
+
 	return ret;
 }
 
-- 
1.7.9.5

